### PR TITLE
use indexedDB instead of sqljs

### LINKED ISSUE #414: avoid using sql in browser mode
sql.js added a lot of complexity. we should just use indexed db in browser

### PR DESCRIPTION
**Pull request by qiweiii**
closes #414

### CODE REVIEW

**Review by ermalkaleci — APPROVED** [review]

File: packages/db/package.json [hunk 1]
```diff
@@ -5,13 +5,13 @@
...
-		"build": "yarn pack-wasm; tsc -p ./tsconfig.json"
+		"build": "tsc -p ./tsconfig.json"
 	},
 	"dependencies": {
 		"@acala-network/chopsticks-core": "workspace:*",
 		"@polkadot/util": "^12.5.1",
 		"@polkadot/wasm-util": "^7.2.2",
+		"idb": "^7.1.1",
 		"localforage": "^1.10.0",
 		"sql.js": "^1.8.0",
```
Comment (ermalkaleci) [inline packages/db/package.json:16]: Remove this

**Review by ermalkaleci — COMMENTED** [review]
These

File: packages/db/package.json [hunk 2]
```diff
@@ -5,13 +5,13 @@
...
 	"scripts": {
 		"clean": "rm -rf lib tsconfig.tsbuildinfo",
-		"pack-wasm": "scripts/pack-wasm.cjs",
-		"build": "yarn pack-wasm; tsc -p ./tsconfig.json"
+		"build": "tsc -p ./tsconfig.json"
 	},
 	"dependencies": {
 		"@acala-network/chopsticks-core": "workspace:*",
 		"@polkadot/util": "^12.5.1",
 		"@polkadot/wasm-util": "^7.2.2",
```
Comment (ermalkaleci) [inline packages/db/package.json:13]: these

File: packages/db/src/browser.ts [hunk 3]
```diff
@@ -1,13 +1,84 @@
...
+
+  async queryBlockByNumber(number: number): Promise<BlockEntry | null> {
+    const db = await this.datasource
+    const block = await db.getFromIndex('block', 'byNumber', number)
+    return block ?? null
+  }
+
+  async queryHighestBlock(): Promise<BlockEntry | null> {
+    const db = await this.datasource
+    const block = await db.getAllFromIndex('block', 'byNumber')
```
Comment (ermalkaleci) [inline packages/db/src/browser.ts:56]: can't we just get one?

Comment (qiweiii) [reply packages/db/src/browser.ts:56]: `getFromIndex` needs to have the third argument

File: packages/db/src/browser.ts [hunk 4]
```diff
@@ -1,13 +1,85 @@
...
+  async saveStorage(blockHash: `0x${string}`, key: `0x${string}`, value: `0x${string}` | null): Promise<void> {
+    const db = await this.datasource
+    await db.put('keyValue', {
+      blockHash,
+      key,
+      value,
+    })
+  }
+
+  async queryStorage(_blockHash: `0x${string}`, key: `0x${string}`): Promise<KeyValueEntry | null> {
```
Comment (ermalkaleci) [inline packages/db/src/browser.ts:80]: this needs to query with blockHash

**Review by ermalkaleci — CHANGES_REQUESTED** [review]
KeyValue needs to be stored and retrieved use both blockHash and storageKey. Easiest way I can think is to combine hash + key

File: packages/db/src/browser.ts [hunk 5]
```diff
@@ -1,13 +1,90 @@
...
+    return db.count('block')
+  }
+
+  async saveStorage(blockHash: `0x${string}`, key: `0x${string}`, value: `0x${string}` | null): Promise<void> {
+    const db = await this.datasource
+    await db.put('keyValue', {
+      keyAndBlockHash: `${key}-${blockHash}`,
+      blockHash,
+      key,
+      value,
```
Comment (ermalkaleci) [inline packages/db/src/browser.ts:81]:
```suggestion
      key: `${blockHash}-${key}`,
      value,
```

File: packages/db/src/browser.ts [hunk 6]
```diff
@@ -1,13 +1,90 @@
...
+}
+export class IdbDatabase implements Database {
+  datasource: Promise<IDBPDatabase<Schema>>
 
   constructor(location: string) {
-    super()
-    this.datasource = openDb(location)
+    this.datasource = openDB<Schema>(location, 1, {
+      upgrade(db) {
+        db.createObjectStore('keyValue', { keyPath: 'keyAndBlockHash' })
```
Comment (ermalkaleci) [inline packages/db/src/browser.ts:25]:
```suggestion
        db.createObjectStore('keyValue', { keyPath: 'key' })
```

File: packages/db/src/browser.ts [hunk 7]
```diff
@@ -1,13 +1,90 @@
...
+interface KeyValue extends KeyValueEntry {
+  keyAndBlockHash: string
+}
 
-export class SqljsDatabase extends BaseSqlDatabase {
-  datasource: Promise<DataSource>
+interface Schema extends DBSchema {
+  keyValue: {
+    key: string
+    value: KeyValue
```
Comment (ermalkaleci) [inline packages/db/src/browser.ts:11]:
```suggestion
    value: string
```

File: packages/db/src/browser.ts [hunk 8]
```diff
@@ -1,13 +1,90 @@
...
+      keyAndBlockHash: `${key}-${blockHash}`,
+      blockHash,
+      key,
+      value,
+    })
+  }
+
+  async queryStorage(blockHash: `0x${string}`, key: `0x${string}`): Promise<KeyValueEntry | null> {
+    const db = await this.datasource
+    const keyValue = await db.get('keyValue', `${key}-${blockHash}`)
```
Comment (ermalkaleci) [inline packages/db/src/browser.ts:87]:
```suggestion
    const keyValue = await db.get('keyValue', `${blockHash}-${key}`)
```

File: packages/web-test/tests/index.spec.ts [hunk 9]
```diff
@@ -30,4 +46,32 @@ test.describe('index', () => {
...
+    const blocksCount = await page.evaluate(() => globalThis.chain.db?.blocksCount())
+    expect(blocksCount).toBe(1)
+    await page.evaluate(
+      (hightestBlock) => globalThis.chain.db?.deleteBlock(hightestBlock?.hash as HexString),
+      hightestBlock,
+    )
+    const blocksCount2 = await page.evaluate(() => globalThis.chain.db?.blocksCount())
+    expect(blocksCount2).toBe(0)
+
+    await page.evaluate(() => globalThis.chain.db?.saveStorage('0x', '0x', null))
```
Comment (ermalkaleci) [inline packages/web-test/tests/index.spec.ts:73]: add some tests with values

File: packages/web-test/tests/index.spec.ts [hunk 10]
```diff
@@ -30,4 +46,32 @@ test.describe('index', () => {
...
+  test('chain indexedDB works', async ({ page }) => {
+    test.setTimeout(5 * 60 * 1000) // 5 minutes timeout
+    // chain is ready
+    await expect(page.locator('#blocks-section')).toHaveText(/4000000/, { timeout: 60_000 })
+    await page.getByText(/build block/i).click()
+    // wait for new block
+    await expect(page.locator('#blocks-section')).toHaveText(/4000001/, { timeout: 200_000 })
+
+    // test db methods
+    const hightestBlock = await page.evaluate(() => globalThis.chain.db?.queryHighestBlock())
```
Comment (ermalkaleci) [inline packages/web-test/tests/index.spec.ts:59]: needs to build more than 1 block to test this

**Review by ermalkaleci — CHANGES_REQUESTED** [review]
We need to reduce storage as much as possible. There's no need to store anything else, just `key: value`

File: packages/db/src/browser.ts [hunk 11]
```diff
@@ -1,13 +1,90 @@
...
+    return keyValue ?? null
```
Comment (ermalkaleci) [inline packages/db/src/browser.ts:88]:
```suggestion
    return keyValue ? { key, value: keyValue.value } : null
```

**Review by ermalkaleci — APPROVED** [review]

**Pull request merged**